bpf/bpf_sock.c

```diff
@@ -382,6 +385,35 @@ static __always_inline int __sock4_xlate_fwd(struct bpf_sock_addr *ctx,
 	if (sock4_skip_xlate(svc, orig_key.address))
 		return -EPERM;
 
+#ifdef ENABLE_L7_LB
+	/* Do not perform service translation at socker layer for
+	 * services with L7 load balancing as we need to postpone
+	 * policy enforcement to take place after l7 load balancer and
+	 * we can't currently do that from the socket layer.
+	 */
+	if (lb4_svc_is_l7loadbalancer(svc)) {
+		/* TC level eBPF datapath does not handle node local traffic,
+		 * but we need to redirect for L7 LB also in that case.
+		 */
+		if (is_defined(BPF_HAVE_NETNS_COOKIE) && in_hostns) {
+			/* Use the L7 LB proxy port as a backend. Normally this
+			 * would cause policy enforcement to be done before the
+			 * L7 LB (which should not be done), but in this case
+			 * (node-local nodeport) there is no policy enforcement
+			 * anyway.
+			 */
+			l7backend.address = bpf_htonl(0x7f000001);
+			l7backend.port = (__be16)svc->l7_lb_proxy_port;
```

Comment:
Is there one of these per nodeport or something? Or, how does the L7LB know which service is being contacted?

Reply:
TL;DR: Typically yes.

My initial design was to try always preserve the service IP in the destination address so that the Envoy listener could make a difference between them if multiple `CiliumEnvoyConfig.Spec.Services` are defined. However, in this case of node-port from the host networking namespace there is no bpf program on the TC level where we could redirect the traffic. And in this case the destination IP is the node address anyway, rather than the service IP.

It looks like doing this in most cases would allow us to simplify the TC level handling, basically getting rid of the per-packet lb stuff there. Given this I'd be inclined to document that the destination IP address as seen by the Envoy Listener can be the local host address, node address, as well as the actual service IP.

At least for K8s Ingress implementation this does not matter as there is only one `CEC.Spec.Services` and each Ingress gets their own Envoy listener in a separate listening port.

Reply:
https://github.com/cilium/cilium/pull/18894/files#diff-2b655ba613ca8386fd9f4c383200480492ee25711e24a575f3fc8b33531b4672R246-R253

Reply:
https://github.com/cilium/cilium/pull/18894/files#diff-2b655ba613ca8386fd9f4c383200480492ee25711e24a575f3fc8b33531b4672R246-R253